pkg/engine/engine.go

```diff
@@ -161,53 +174,98 @@ func (e *Engine) DoBuild(ctx context.Context, testplan string, builder string, i
 		return nil, fmt.Errorf("unrecognized builder: %s", builder)
 	}
 
-	// Compile all configurations to coalesce. Precedence (highest to lowest):
+	// This var compiles all configurations to coalesce.
+	//
+	// Precedence (highest to lowest):
+	//
 	//  1. CLI --run-param, --build-param flags.
 	//  2. .env.toml.
 	//  3. Test plan definition.
-	//  4. Builder defaults (applied by the builder).
-	var cfgs []map[string]interface{}
+	//  4. Builder defaults (applied by the builder itself, nothing to do here).
+	//
+	var cfg config.CoalescedConfig
 
-	// 3. Get the base configuration of the build strategy.
-	planCfg, ok := plan.BuildStrategies[builder]
-	if !ok {
+	// 3. Add the base configuration of the build strategy.
+	if c, ok := plan.BuildStrategies[builder]; !ok {
 		return nil, fmt.Errorf("test plan does not support builder: %s", builder)
+	} else {
+		cfg = cfg.Append(c)
 	}
-	cfgs = append(cfgs, planCfg)
 
 	// 2. Get the env config for the builder.
-	envCfg, ok := e.envcfg.BuildStrategies[builder]
-	if ok {
-		cfgs = append(cfgs, envCfg)
-	}
+	cfg = cfg.Append(e.envcfg.BuildStrategies[builder])
 
 	// 1. Get overrides from the CLI.
-	if overrides := input.BuildConfig; overrides != nil {
-		// We have config overrides.
-		cfgs = append(cfgs, overrides.(map[string]interface{}))
-	}
+	cfg = cfg.Append(comp.Global.BuildConfig)
 
-	// Coalesce all configs and deserialize into the provided type.
-	cfg, err := coalesceConfigsIntoType(bm.ConfigType(), cfgs...)
+	// Coalesce all configurations and deserialise into the config type
+	// mandated by the builder.
+	obj, err := cfg.CoalesceIntoType(bm.ConfigType())
 	if err != nil {
 		return nil, fmt.Errorf("error while coalescing configuration values: %w", err)
 	}
 
-	input.BuildID = uuid.New().String()
-	input.Directories = e.envcfg
-	input.TestPlan = plan
-	input.BuildConfig = cfg
-	input.EnvConfig = *e.envcfg
+	var (
+		// no need to synchronise access, as each goroutine will write its
+		// response in its index.
+		ress   = make([]*api.BuildOutput, len(comp.Groups))
+		errgrp = errgroup.Group{}
+		cancel context.CancelFunc
+	)
+
+	// obtain an explicitly cancellable context so we can stop build jobs if
+	// something fails.
+	ctx, cancel = context.WithCancel(ctx)
+	defer cancel()
+
+	// Trigger a build for each group, and wait until all of them are done.
+	for i, grp := range comp.Groups {
+		ii, grpp := i, grp // captures
```

Comment:
No need to introduce new names, it is a common pattern to just do `i := i` when you need to capture the var.

Reply:
Done.